src/shims/posix/linux/sync.rs

```diff
@@ -0,0 +1,88 @@
+use crate::*;
+use rustc_target::abi::{Align, Size};
+
+/// Implementation of the SYS_futex syscall.
+pub fn futex<'tcx>(
+    this: &mut MiriEvalContext<'_, 'tcx>,
+    args: &[OpTy<'tcx, Tag>],
+    dest: PlaceTy<'tcx, Tag>,
+) -> InterpResult<'tcx> {
+    // The amount of arguments used depends on the type of futex operation.
+    // Some users always pass all arguments, even the unused ones, due to how they wrap this syscall in their code base.
+    // Some other users pass only the arguments the operation actually needs. So we don't use `check_arg_count` here.
+    if !(4..=7).contains(&args.len()) {
+        throw_ub_format!("incorrect number of arguments for futex syscall: got {}, expected between 4 and 7 (inclusive)", args.len());
+    }
+
+    // The first three arguments (after the syscall number itself) are the same to all futex operations:
+    //     (int *addr, int op, int val).
+    // Although note that the first one is often passed as a different pointer type, e.g. `*const AtomicU32` or `*mut u32`.
+    let addr = this.deref_operand(args[1])?;
```

Comment:
Wait, didn't you say that this point may in fact be dangling? That won't work like this. `deref_operand` requires that the pointer be dereferncable for the given type.

I am okay leaving this as a FIXME though. But eventually we should have a test for this.

Reply:
> Wait, didn't you say that this point may in fact be dangling?

I did! I (mis)interpreted your comment above ..

> Then later you can use `this.read_scalar(futex_val_place.offset)` to read the actual data, and `futex_val_place.ptr.assert_ptr()` to get the `Pointer`. (`deref_operand` does the necessary int-to-ptr conversion.)

.. as that only `read_scalar()` requires the address to be readable, and that `deref_operand` only does the conversion to a pointer.


Anyway, this was the reason I kept the original `addr` around as a `TyOp`, separate from the `Pointer`, such that the FUTEX_WAIT path can still call `deref_operand` (or `read_scalar_at_offset`, which already does `deref_operand`) on it.

What's a good way to do this? Both operations need `addr` as `Pointer`, but one also needs to dereference the pointer (*as i32*).

Reply:
I'm using `force_ptr` now instead of `deref_operand`. But using that `Pointer` to read the `i32` pointee in FUTEX_WAIT seems tricky.

Now I'm using `check_ptr_access` followed by `this.memory.get_raw(addr.alloc_id)?.read_scalar`, but I don't know if that's a proper way to do it.

Reply:
Yes this is a tricky situation which I do not think we've had before. But `get_raw` feels like going down way too far the stack of abstractions.

I think the best way is to delay part of the decision: here at the top, just do `let futex_val = this.read_immediate`. Then also get the ptr value via `force_ptr(futex_val.to_scalar()?)`. In the `futex_wait` branch, do `deref_operand(futex_val.into())?`. At least I think that should work.

Reply:
That only works if I do:

```rust
this.read_scalar(this.deref_operand(addr.into())?.offset(Size::ZERO, MemPlaceMeta::None, this.machine.layouts.i32, this)?.into())?.to_i32()?
```

or

```rust
this.read_scalar_at_offset(addr.into(), 0, this.machine.layouts.i32)?.to_i32()?
```

in the futex_wait branch. Otherwise it refuses to read a `AtomicI32` as `i32` again.

Reply:
Ah right.  Well then `read_scalar_at_offset` is probably indeed the best version. Sorry for the detour, I didn't realize the tricky "only sometimes deref" semantics here that we didn't have before in Miri.

Reply:
Hm, now it fails if somebody calls syscall() with that futex address as a `usize`. Neither parking_lot nor std do that, but it might make sense as in Rust `*const i32` is not `Send`, while `usize` is. And since this pointer may dangle, sending it to another thread might make sense.

I ran into this while making a test:
```rust
    let futex = 0;
    let futex_ptr = &futex as *const i32 as usize; // usize, otherwise we get an error about *const i32 not being Send.

    let t2 = thread::spawn(move || {
        thread::sleep(Duration::from_millis(200));
        unsafe {
            assert_eq!(libc::syscall(
                libc::SYS_futex,
                futex_ptr, // Pointer to the local variable above. It's okay if this pointer dangles here.
                libc::FUTEX_WAKE,
                10, // Wake up at most 10 threads.
            ), 1); // Woken up one thread.
        }
    });

    unsafe {
        assert_eq!(libc::syscall(
            libc::SYS_futex,
            futex_ptr,
            libc::FUTEX_WAIT,
            0,
            ptr::null::<libc::timespec>(),
        ), 0);
    }
```

It'd definitely be nicer to properly pass a pointer instead, but this code is valid.

So uh, should we care about this edge case? If so, do we go back to `memory.get_raw`, or is there a better way?

(It doesn't matter for the issues this PR is trying to solve.)

(I'm starting to dislike C's variadic functions more every second. :upside_down_face:)

Reply:
Ah, pointers are annoying...
I'd say add a FIXME. We usually rely on the program to set a reasonable prototype for the function even if, technically, many other prototypes would also work; I think it makes sense to also do that for variadics to the extend that we can get away with.

Reply:
Yeah, I think it's fine to leave it as is then. Updated the `FIXME` at `syscall` to also mention this.

Reply:
> Hm, now it fails if somebody calls syscall() with that futex address as a usize. Neither parking_lot nor std do that, but it might make sense as in Rust *const i32 is not Send, while usize is. And since this pointer may dangle, sending it to another thread might make sense.

I am confused actually, why does this fail? Without `deref_operand` I think this should work, but I might forget some check somewhere.

Reply:
Or was the actual problem that you passed `0usize`? In that case the value would have been the problem, not the type -- i.e., `0usize as *const ()` would not work either.

Reply:
> I am confused actually, why does this fail?

Because in the example I also used the `usize` for the wait operation, which does dereference it.

Reply:
But it doesn't do that with `deref_operand`. So as long as the pointer points to something valid, with your current code, I think it should work fine both with usize and with a pointer type.

Reply:
Oh, `read_scalar_at_offset` uses `deref_operand` internally...